### PR TITLE
Support for publishing multiple versions

### PR DESCRIPTION
**Pull request by mbhealy**
### Summary
This PR adds support for publishing multiple stable versions (and a single live version) of the spec. A preview of the result is currently available at: https://mbhealy.github.io/intro.html

### Details and comments
The title of the project is updated with the current version using an environment variable in the sphinx build process.
The list of links to the available versions is built out of all branches named 'stable/**'
I'm using a new bash script to build each version in a loop and copy the html files to a subdirectory, that is linked to in the sidebar. I tried multiple other approaches but all were very wonky. Getting this to work with Sphinx was a real pain. I'm open to suggestions.

Needs to be updated before merge with relevant constants for the openqasm site instead of my personal github pages repo. My repo is a bit weird because I haven't updated the default branch to main from master.

Closes https://github.com/openqasm/openqasm/issues/330

### CODE REVIEW

**Comment by mbhealy** [comment]
The Live link the sidebar is hardcoded to [openqasm.github.io](http://openqasm.github.io/), so be careful which URL you're on

File: source/conf.py [hunk 1]
```diff
@@ -17,8 +17,10 @@
 # -- Project information -----------------------------------------------------
 from typing import List
 
-project = 'OpenQASM Live Specification'
-copyright = '2017-2020, Andrew W. Cross, Lev S. Bishop, John A. Smolin, Jay M. Gambetta'
+version = os.getenv('VERSION','Live')
+
+project = f'OpenQASM {version} Specification'
+copyright = '2017-2023, Andrew W. Cross, Lev S. Bishop, John A. Smolin, Jay M. Gambetta'
```
Comment (blakejohnson) [inline source/conf.py:23]: Not a new problem introduced here, but at some point we need to investigate the proper way to claim copyright to the OpenQASM community, Qiskit, or whatever entity makes sense from a legal point of view.

File: source/conf.py [hunk 2]
```diff
@@ -59,6 +61,21 @@
...
 
+version_list_var = os.getenv('VERSION_LIST')
+extra_nav_links = {'Live Version': '/index.html'} # default link to Live version
+
+if version_list_var is not None:
+    version_list = version_list_var.split(',')
+    for ver in version_list:
+        extra_nav_links[f'Version {ver}'] = f'/versions/{ver}/index.html'
+
+print(extra_nav_links)
```
Comment (blakejohnson) [inline source/conf.py:72]: Is this just for debugging?

File: source/conf.py [hunk 3]
```diff
@@ -76,3 +93,10 @@
 numfig = True
 # Necessary setting for sphinxcontrib-bibtex >= 2.0.0
 bibtex_bibfiles = ['bibliography.bib']
+
+variables_to_export = [
+    "version",
+]
+frozen_locals = dict(locals())
+rst_epilog = '\n'.join(map(lambda x: f".. |{x}| replace:: {frozen_locals[x]}", variables_to_export))
```
Comment (blakejohnson) [inline source/conf.py:101]: Please add a comment about where/how this value is used by Sphinx.

File: .github/workflows/publish-all-verions.yml [hunk 4]
```diff
@@ -22,12 +21,13 @@ jobs:
...
+          ./buildPublishable.sh mbh-versioning ./publish_build remotes/origin
 
       - name: Deploy 🚀
         uses: JamesIves/github-pages-deploy-action@v4.3.3
         with:
           token: ${{ secrets.GHPAGES_DEPLOY_TOKEN }}
-          repository-name: openqasm/openqasm.github.io
-          branch: main
-          folder: build/html
+          repository-name: mbhealy/mbhealy.github.io
```
Comment (blakejohnson) [inline .github/workflows/publish-all-verions.yml:31]: I assume this is what you will update before merging?

File: source/conf.py (hunk 2 above)
Comment (mbhealy) [reply source/conf.py:72]: Yes, I think it's useful to print this for inclusion in the CI logs.

File: .github/workflows/publish-all-verions.yml (hunk 4 above)
Comment (mbhealy) [reply .github/workflows/publish-all-verions.yml:31]: Yes, will be updated before merge.

File: source/conf.py (hunk 3 above)
Comment (mbhealy) [reply source/conf.py:106]: Added in https://github.com/openqasm/openqasm/pull/452/commits/517cbc386d5d70ebce6ebbce1e7a419e91e65dee

File: .github/workflows/publish-all-verions.yml (hunk 4 above)
Comment (mbhealy) [reply .github/workflows/publish-all-verions.yml]: Done in https://github.com/openqasm/openqasm/pull/452/commits/0d676cfb3e7eb5b80ee985a4d5a7aacecc7288e8

File: source/conf.py (hunk 1 above)
Comment (mbhealy) [reply source/conf.py:23]: Added a new issue here: https://github.com/openqasm/openqasm/issues/466

File: buildPublishable.sh [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+  then
+    versionList="${versionNum}"
+  else
+    versionList="${versionList},${versionNum}"
+  fi
+done
+
+# Now build each stable version and copy to destination folder
+echo "VersionList is ${versionList}"
+for branch in `git for-each-ref --format='%(refname:short)' --sort=-refname refs/${repoPath}/stable/`; do
```
Comment (blakejohnson) [inline buildPublishable.sh:45]: Why not iterate over `versionList` instead?

**Review by blakejohnson — APPROVED** [review]

**Comment by blakejohnson** [comment]
Sorry for the delay, @mbhealy. This looks good to me.

**Pull request merged**